### DISCUSSION TITLE
Prune / Clean points to save space

### DISCUSSION THREAD

**Discussion by MrColumbo** [Freika/dawarich#340] [Ideas]
The space used by dawarich increased quite a lot recently and I was wondering if there is a way to automatically remove points where they do not provide any value. For example when I spend 2 hours in a restaurant I like to keep the point when I got there and the point when I left. Depending on the settings on the client dawarich might not even receive any points when I do not move around but it might be nice to be able to bulk delete points for example when there are several points with the same coordinates but different time stamps. In such cases I like to keep the first point and the last point with the same (or very close distance) coordinates.

**Comment by Freika** [comment]
What takes space is reverse geocoding data being gathered for each of your points (assuming you have it enabled). Points themselves take up quite a little space. Your solution might be to disable reverse geocoding if you're trying to save some space. For now, I have no plans to introduce automatic points pruning.

**Comment by MrColumbo** [comment]
thanks for the feedback. In my case it was a json log file which kept growing a lot due to of error messages

**Comment by Freika** [comment]
Have a look at the docker compose file in the repo, there are options to limit amount of logs via log rotation

**Comment by willemsFEB** [comment]
I know this is an ancient discussion, but for anyone arriving at this page who want to do something similar, you can do this yourself with a script.

I made a script that prunes points within my areas (as this is usually where I spend most time). It's a basic greedy algorithm that removes points as long as the point before and after are within a 1 hour period. It runs on all points from yesterday 23h until today 23h, and then I run a cronjob that runs this daily at 23h30 (so before the reverse geocoding job runs at midnight).

``` ruby
user = User.find_by(email: "your@domain.com")

# Define time range: yesterday 23:00 to today 23:00
start_time = Time.current.yesterday.change(hour: 23, min: 0, sec: 0)
end_time = Time.current.change(hour: 23, min: 0, sec: 0)

puts "Processing all areas from #{start_time} to #{end_time}"
puts "=" * 50

# Iterate through all user's areas
user.areas.find_each do |area|
  puts "\nArea: #{area.name} (center: #{area.latitude}, #{area.longitude}, radius: #{area.radius}m)"
  
  # Get points in this area within the time window
  points = user.points.where(
    timestamp: start_time..end_time
  ).where(
    "ST_DWithin(lonlat, :center, :radius)",
    center: "POINT(#{area.longitude} #{area.latitude})",
    radius: area.radius
  ).order(:timestamp)
  
  if points.empty?
    puts "  No points found in this area."
    next
  end
  
  # Prune points to 1-hour intervals
  sorted = points.to_a
  kept = [sorted.first]
  
  sorted.each_cons(2) do |prev, curr|
    kept << prev if curr.timestamp - kept.last.timestamp > 1.hour
  end
  kept << sorted.last unless kept.include?(sorted.last)
  
  to_delete = sorted - kept
  to_delete.each(&:destroy)
  
  max_gap = kept.each_cons(2).map { |a,b| b.timestamp - a.timestamp }.max || 0
  
  puts "  Total: #{sorted.count}, Kept: #{kept.count}, Removed: #{to_delete.count}"
end

puts "\n" + "=" * 50
puts "Done processing all areas."
```

you then mount this script to the dawarich_app image, (eg at /scripts/prune_data.rb), and then run a cronjob like:

```
30 23 * * * /usr/bin/docker exec -i dawarich_app bin/rails runner /scripts/prune_data.rb >> ~/logs/dawarich_prune.log 2>&1
```

**Comment by Freika** [comment]
Returning to an old topic, in the next releases there will be db changes and also instructions on how to offload raw data stored in the db to external archives

**Comment by jwilleke** [comment]
Adding a data point from a large self-hosted instance, because this thread is about **disk** space and there is a **memory** failure mode next to it that I think is under-appreciated — and largely invisible.

Also, @Freika, your 2024 answer here still holds up: reverse geocoding data is indeed the bulk of it. We have `STORE_GEODATA=true` and it shows.

**Setup:** Dawarich 1.10.0, `postgis/postgis:17-3.5-alpine`, single-node k3s, ~27k points/day since 2026-07-16.

`points` is now **6.5 GB** — heap 3291 MB plus indexes 3173 MB, 2.4M rows. The part that caught us out is that this exceeds the **4 GiB memory limit** on the postgres container serving it. Once it crossed that line we started taking cgroup OOM kills: **68 in seven days**, 57 postgres and 11 sidekiq.

**Why this is worth flagging separately:** the postgres kills leave no trace anywhere an operator normally looks. Postgres forks a backend per connection, the cgroup killer takes the fattest backend, and the postmaster reaps it — PID 1 never dies. So the container never restarts, Kubernetes never records `OOMKilled`, and `kubectl get pods` reads `1/1 Running` the whole time. We only found it by reading the host kernel log. Anyone self-hosting at this size could be losing queries with no signal at all.

Postgres is modestly configured (`shared_buffers` 128 MB, `work_mem` 4 MB, `max_connections` 100), so the static footprint is under 1 GiB. The kills show individual backends at **3.5 GB anonymous RSS**, i.e. one query allocating far past `work_mem`. They cluster 06:00-11:00 local — 45 of 68 — which points at a scheduled job rather than user traffic. I have not captured the query yet; if it would be useful I can set `log_min_duration_statement` and report back with the plan.

For context on why "just raise the limit" stopped being satisfying: we have already raised limits five times (db 768Mi → 2Gi → 4Gi, sidekiq 512Mi → 3Gi → 6Gi, app 1536Mi → 3Gi). At ~60 MB/day of growth each raise buys a few weeks.

I saw the note above about upcoming DB changes and instructions for offloading raw data to external archives — that sounds like it addresses the disk half well. The thing that would help the memory half is smaller: **a documented relationship between points volume and required postgres container memory.** Right now hardware requirements suggest 2 GB minimum, which is a reasonable starting point but gives no way to project what a multi-million-point history needs. Even a rough "expect N GB of DB per 100k points, size postgres to at least X" would let operators size for it instead of discovering it via the kernel log.

Two smaller notes for anyone else arriving here:

- `DELETE /api/v1/points/bulk_destroy` does work on self-hosted (it is not Pro-gated there), but it takes explicit `point_ids`, so it is a delete primitive rather than a retention policy — you still have to select the IDs yourself, as @willemsFEB's script does.
- If you run postgres with a memory limit, it is worth checking your host's kernel log for `Memory cgroup out of memory` rather than trusting container restart counts. Ours said 2 restarts while 57 kills had happened.
